Exception is not being raised after adding exception to 'throws' in the IDL

I am having a strange error where the below test fails with InvalidArgument in the throws section in the IDL below, BUT it passes if I remove it. I have confirmed that the exception is being thrown when I step through the code.

Any ideas why this exception would be swallowed?

Thrift version: 0.9.3
I am loading the thrift spec with thrift.load()

===IDL===
```
service DataAggregationService 
{
  ...
  void tripCreatedOrChanged(
    1: TripCreatedOrChangedRequest request
  ) throws (
    1: ServiceError serviceError,
    2: InvalidArgument invalidArgument
  )
}
```

===Handler===
```
@stats.wrapper("servicename.handlers.tripCreatedOrChanged")
@tchannel.thrift.register(servicename.DataAggregationService)
@thrift_handler_wrapper
@coroutine
def tripCreatedOrChanged(request):
    trip_uuid = validate_uuid(request.body.tripUuid)
    yield DataAggregationService.get_and_log_profile_trip_changes(trip_uuid)
```
NOTE: `validate_uuid` raises an InvalidArgument exception if the uuid is not a valid UUID.

===Test===
```
 @mock.patch.object(DataAggregationService, 'get_and_log_profile_trip_changes')
    @gen_test
    def test_trip_created_or_changed_handles_invalid_uuid(self, mock_get_log):
        """Ensure tripCreatedOrChanged raises an InvalidArgument exception if uuid is invalid."""
        request = MockRequest(tripUuid='invalid-uuid')
        with self.assertRaises(InvalidArgument):
            yield data_aggregation.tripCreatedOrChanged(request)

```

Stacktrace:
```
_______________________________________________ TestDataAggregationHandler.test_user_profile_created_or_changed_handles_invalid_uuid _______________________________________________
env/local/lib/python2.7/site-packages/tornado/testing.py:525: in post_coroutine
    timeout=timeout)
env/local/lib/python2.7/site-packages/tornado/ioloop.py:453: in run_sync
    return future_cell[0].result()
env/local/lib/python2.7/site-packages/tornado/concurrent.py:232: in result
    raise_exc_info(self._exc_info)
env/local/lib/python2.7/site-packages/tornado/gen.py:1017: in run
    yielded = self.gen.send(value)
tests/handlers/test_data_aggregation.py:49: in test_user_profile_created_or_changed_handles_invalid_uuid
    yield data_aggregation.userProfileCreatedOrChanged(request)
E   AssertionError: InvalidArgument not raised
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -9,8 +9,10 @@ Changes by Version
   ``tchannel.sync.thrift.client_for`` as planned in 0.18.
 - **BREAKING** - Removed ``tchannel.thrift_request_builder`` as
   planned in 0.18.
-- **BREAKING** - ``tchannel.thrift.register`` returns the wrapped function
-  as-is. This allows calling the function diretly for unit tests.
+- **BREAKING** - Support unit testing endpoints by calling the handler
+  functions directly. This is enabled by changing ``tchannel.thrift.register``
+  to return the registered function unmodified. See Upgrade Guide for more
+  details.
 - Reduced Zipkin submission failures to warnings.
 - Limit the size of arg1 to 16KB.
 - Fix bug which prevented requests from being retried if the candidate

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -9,6 +9,8 @@ Changes by Version
   ``tchannel.sync.thrift.client_for`` as planned in 0.18.
 - **BREAKING** - Removed ``tchannel.thrift_request_builder`` as
   planned in 0.18.
+- **BREAKING** - ``tchannel.thrift.register`` returns the wrapped function
+  as-is. This allows calling the function diretly for unit tests.
 - Reduced Zipkin submission failures to warnings.
 - Limit the size of arg1 to 16KB.
 - Fix bug which prevented requests from being retried if the candidate

--- a/UPGRADE.rst
+++ b/UPGRADE.rst
@@ -6,6 +6,54 @@ Upgrade Guide
 Migrating to a version of TChannel with breaking changes? This guide documents
 what broke and how to safely migrate to newer versions.
 
+From 0.20 to 0.21
+-----------------
+
+- ``tchannel.thrift.register`` returns the original function as-is instead of
+  the wrapped version. This allows writing unit tests that call the handler
+  function directly.
+
+  Previously, if you used the ``tchannel.thrift.register`` decorator to
+  register a Thrift endpoint and then called that function directly from a
+  test, it would return a ``Response`` object if the call succeeded or
+  failed with an expected exception (defined in the Thrift IDL). For example,
+
+  .. code-block:: python
+
+      # service KeyValue {
+      #   string getValue(1: string key)
+      #      throws (1: ValidationError invalid)
+      # }
+
+      @tchannel.thrift.register(kv.KeyValue)
+      def getValue(request):
+          key = request.body.key
+          if key == 'invalid':
+              raise kv.ValidationError()
+          result = # ...
+          return result
+
+      response = getValue(make_request(key='invalid'))
+      if response.body.invalid:
+          # ...
+      else:
+          result = response.body.success
+
+  With 0.21, we have changed ``tchannel.thrift.register`` to return the
+  unmodified function so that you can call it directly and it will behave
+  as expected.
+
+  .. code-block:: python
+
+      @tchannel.thrift.register(kv.KeyValue)
+      def getValue(request):
+          # ...
+
+      try:
+          result = getValue(make_request(key='invalid'))
+      except kv.ValidationError:
+          # ...
+
 From 0.19 to 0.20
 -----------------
 

--- a/tchannel/thrift/rw.py
+++ b/tchannel/thrift/rw.py
@@ -307,10 +307,9 @@ def register(dispatcher, service, handler=None, method=None):
         )
         assert not function.oneway
 
-        handler = build_handler(function, handler)
         dispatcher.register(
             function.endpoint,
-            handler,
+            build_handler(function, handler),
             ThriftSerializer(service._module, function._request_cls),
             ThriftSerializer(service._module, function._response_cls),
         )

--- a/tests/schemes/test_thrift.py
+++ b/tests/schemes/test_thrift.py
@@ -1045,7 +1045,7 @@ def test_exception_status_code_is_set(server, ThriftTest, server_ttypes):
 # as if they weren't registered at all
 
 
-def make_request(**kwargs):
+def body(**kwargs):
     """Constructs fake Request objects.
 
     The ``_headers`` kwarg may be used to set the header on the Request
@@ -1053,10 +1053,7 @@ def make_request(**kwargs):
     """
     request = mock.Mock()
     for k, v in kwargs.items():
-        if k == '_headers':
-            request.headers = v
-        else:
-            setattr(request.body, k, v)
+        setattr(request, k, v)
     return request
 
 
@@ -1066,7 +1063,7 @@ def test_void_call_directly(server, ThriftTest):
     def testVoid(request):
         pass
 
-    resp = testVoid(make_request())
+    resp = testVoid(Request())
     assert resp is None
 
 
@@ -1077,7 +1074,7 @@ def test_void_with_headers_call_directly(server, ThriftTest):
         assert request.headers == {'req': 'header'}
         return Response(headers={'resp': 'header'})
 
-    resp = testVoid(make_request(_headers={'req': 'header'}))
+    resp = testVoid(Request(headers={'req': 'header'}))
     assert resp.headers == {'resp': 'header'}
     assert resp.body is None
 
@@ -1088,7 +1085,7 @@ def test_non_void_call_directly(server, ThriftTest):
     def testString(request):
         return request.body.thing
 
-    resp = testString(make_request(thing='howdy'))
+    resp = testString(Request(body=body(thing='howdy')))
     assert resp == 'howdy'
 
 
@@ -1109,9 +1106,9 @@ def test_non_void_with_headers_cal_directly(
         )
 
     resp = testStruct(
-        make_request(
-            _headers={'req': 'header'},
-            thing=client_ttypes.Xtruct("req string"),
+        Request(
+            headers={'req': 'header'},
+            body=body(thing=client_ttypes.Xtruct("req string")),
         )
     )
 

--- a/tests/schemes/test_thrift.py
+++ b/tests/schemes/test_thrift.py
@@ -23,6 +23,7 @@ from __future__ import division
 from __future__ import print_function
 from __future__ import unicode_literals
 
+import mock
 import pytest
 from tornado import gen
 
@@ -1037,3 +1038,83 @@ def test_exception_status_code_is_set(server, ThriftTest, server_ttypes):
     )
 
     assert 1 == res.status_code
+
+
+##############################################################################
+# Calling registered functions directly should be equivalent to calling them
+# as if they weren't registered at all
+
+
+def make_request(**kwargs):
+    """Constructs fake Request objects.
+
+    The ``_headers`` kwarg may be used to set the header on the Request
+    object.
+    """
+    request = mock.Mock()
+    for k, v in kwargs.items():
+        if k == '_headers':
+            request.headers = v
+        else:
+            setattr(request.body, k, v)
+    return request
+
+
+def test_void_call_directly(server, ThriftTest):
+
+    @server.thrift.register(ThriftTest)
+    def testVoid(request):
+        pass
+
+    resp = testVoid(make_request())
+    assert resp is None
+
+
+def test_void_with_headers_call_directly(server, ThriftTest):
+
+    @server.thrift.register(ThriftTest)
+    def testVoid(request):
+        assert request.headers == {'req': 'header'}
+        return Response(headers={'resp': 'header'})
+
+    resp = testVoid(make_request(_headers={'req': 'header'}))
+    assert resp.headers == {'resp': 'header'}
+    assert resp.body is None
+
+
+def test_non_void_call_directly(server, ThriftTest):
+
+    @server.thrift.register(ThriftTest)
+    def testString(request):
+        return request.body.thing
+
+    resp = testString(make_request(thing='howdy'))
+    assert resp == 'howdy'
+
+
+def test_non_void_with_headers_cal_directly(
+    server, service, ThriftTest, server_ttypes, client_ttypes
+):
+
+    # Given this test server:
+
+    @server.thrift.register(ThriftTest)
+    def testStruct(request):
+        assert request.headers == {'req': 'header'}
+        assert request.body.thing.string_thing == 'req string'
+
+        return Response(
+            server_ttypes.Xtruct(string_thing="resp string"),
+            headers={'resp': 'header'},
+        )
+
+    resp = testStruct(
+        make_request(
+            _headers={'req': 'header'},
+            thing=client_ttypes.Xtruct("req string"),
+        )
+    )
+
+    assert isinstance(resp, Response)
+    assert resp.headers == {'resp': 'header'}
+    assert resp.body == client_ttypes.Xtruct("resp string")

